Automattic/wp-calypso
# Staging sites: Prevent navigation on upgrades and inbox menu pages.

Related to https://github.com/Automattic/dotcom-forge/issues/1881

## Proposed Changes

This pr introduces a new router middleware to prevent the navigation on all the pages under "upgrades" and "inbox" menus for staging sites.

It prevents access to all `/domains*`, `/plans*`, `/purchases*`, `/email*` and `/inbox*` sites.

## Testing Instructions



All pages mentioned below follow the following pattern:

`calypso.localhost/<pattern>/<site URL>`

## Upgrades menu

Includes the following pages:

### Plans

- `/plans`
- `/plans/compare`
- `/plans/compare/:domain`
- `/plans/features`
- `/plans/features/:domain`
- `/plans/features/:feature/:domain`
- `/plans/my-plan`
- `/plans/my-plan/trial-upgraded/:domain`
- `/plans/my-plan/:site`
- `/plans/select/:plan/:domain`
- `/plans/storage/:site`
- `/plans/upsell/:product/:site`
- `/plans/:intervalType?/:site`
- `/plans/jetpack-free/welcome`,
- `/plans/jetpack-boost/welcome`,
- `/plans/jetpack-social/welcome`,
- `/plans/complete/:site/:lang`,
- `/plans/:duration/:site`

All plan endpoints use a common `commonHandler` array, so test one, tests them all (excluded: `/plans/features/:feature/:domain` and `/plans/my-plan/trial-upgraded/:domain` where I added the middleware in there as well)

### Addons

- `/add-ons`
- `/add-ons/:site`

All add-ons endpoints use a common `commonHandler` array, so test one, tests them all.

### Domains

- All `/domains` pages 

all pages are registered via `registerStandardDomainManagementPages`, and the middleware is there, so test one of them tests them all 

### Purchases

- `/purchases`
- ` /purchases/subscriptions`
- `/purchases/:site`
- `/purchases/subscriptions/:site`
- `/purchases/subscriptions/:site/:purchaseId`
- `/purchases/subscriptions/:site/:purchaseId/cancel`
- `/purchases/subscriptions/:site/:purchaseId/confirm-cancel-domain`
- `/purchases/subscriptions/:site/:purchaseId/payment-method/add`
- `/purchases/subscriptions/:site/:purchaseId/payment-method/change/:cardId`
- `/purchases/add-payment-method/:site`
- `/purchases/billing-history/:site`
- `/purchases/billing-history/:site/:receiptId`
- `/purchases/payment-methods/:site`

All purchase endpoints use a common `commonHandler` array, so test one, tests them all.

### Email
- All `/email` endpoints
- All `/inbox` endpoints

All email endpoints use a common `commonHandler` array, so test one, tests them all.

## Inbox menu

### Email
- All `/email` endpoints
- All `/inbox` endpoints

All email endpoints use a common `commonHandler` array, so test one, tests them all.

### Test page redirection for staging sites
1. Create a staging site
2.  Try to navigate to `wordpress.com/<pattern>/<your site>`
3. Ensure you are being redirected back to the `/home` page.

### Test pages for non-staging sites
1. Create a simple site
2.  Try to navigate to `wordpress.com/<pattern>/<your site>`
3. Ensure you are  NOT being redirected back to the `/home` page.

## Pre-merge Checklist
